src/Blorc.OpenIdConnect/Extensions/ObjectExtensions.cs

```diff
@@ -0,0 +1,79 @@
+﻿namespace Blorc.OpenIdConnect
+{
+    using System.Collections;
+    using System.Collections.Generic;
+    using System.Reflection;
+    using System.Security.Claims;
+    using Blorc.Reflection;
+
+    public static class ObjectExtensions
+    {
+        public static IEnumerable<Claim> AsClaims(this object instance, string claimType = null)
+        {
+            if (instance is null)
+            {
+                yield break;
+            }
+
+            var instanceType = instance.GetType();
+            if (instanceType.IsPrimitiveEx())
+            {
+                if (string.IsNullOrWhiteSpace(claimType))
+                {
+                    yield break;
+                }
+
+                yield return new Claim(claimType, instance?.ToString() ?? string.Empty);
+            }
+            else  if (instanceType.IsCollection() && instance is IEnumerable items)
+            {
+                foreach (var item in items)
+                {
+                    foreach (var claim in item.AsClaims())
+                    {
+                        yield return claim;
+                    }
+                }
+            }
+            else
+            {
+                var propertyInfos = instanceType.GetProperties(BindingFlags.Instance | BindingFlags.Public);
+                foreach (var propertyInfo in propertyInfos)
+                {
+                    var propertyValue = propertyInfo.GetValue(instance);
+                    if (propertyValue is null)
+                    {
+                        continue;
+                    }
+
+                    var propertyValueType = propertyValue.GetType();
+                    var claimTypeAttribute = propertyInfo.GetCustomAttribute<ClaimTypeAttribute>();
```

Comment:
Instead of using this collection / primitive check, why not immediately call this method recursively?

```
if (propertyValue is not null)
{
    foreach (var claim in propertyValue.AsClaims(claimTypeAttribute.ClaimType))
    {
        yield return claim;
    }
}
```

Technically we could even ignore the null check (but I think it's good to add it anyway).

Reply:
Will check, but probably an scenario that I don't remember well in this moment ;)